fsharp.extraction.targets

```diff
@@ -14,7 +14,7 @@
                 3. fsti files are filtered out
                 4. the order of files is unchanged.
     -->
-    <UsingTask TaskName="TransformFStarSourceFiles" TaskFactory="CodeTaskFactory" AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.v4.0.dll" >
+    <UsingTask TaskName="TransformFStarSourceFiles" TaskFactory="RoslynCodeTaskFactory" AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.Core.dll">
```

Comment:
You might want to do it this way to keep this backwards compatible: https://github.com/FStarLang/FStar/pull/2308/files#diff-07d2f3ff15bc554e5da8657aa18e55dd9ded3b8550a057916f4a0ec1054ddd6e

Reply:
Are you sure this is necessary? I think latest MSBuild has `RoslynCodeTaskFactory` task too. Maybe that change needed for `xbuild`? I test build with Visual Studio 2022, but did not test execution from command line. Let me check.

Reply:
Exactly, this was needed in a situation where only older msbuild was available (e.g. older vs installed or older build tools on a ci machine)..

Reply:
According to https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-roslyncodetaskfactory this task supported starting from VS 2017 which is pretty "old". VS2015 at this time on extended maintainance which is debatable target. Do you or core team really need to support this environment?
If yes, can you please tell me what configuation it is, since that's relatively ineresting, and I would like to test this then, if that's has to be supported.

Reply:
Well, I can confirm it was a real issue for me in my previous role. 
Also, I'm not sure if F* itself has already upgraded from VS2015, I don't think that was the case a year ago.

Reply:
That's quiete depressing then. Year ago that should not be an issue either. Should I ask @nikswamy about what versions should be supported?

Reply:
We use VS 2019 community edition.

That said, I am revising the use of F# in the FStar repo. It will no longer be the case that the source files build directly in F# and I am also considering removing this fsharp.extraction.targets file entirely, since (from my other comment) I can't seem to get this build task to work on Linux with a native fstar executable.

Reply:
What do you use for dotnet on Linux? mono? Can you consider using dotnet? I’m positive that this fix plus sibling PR make it work. I’m positive that I can make it work on Linux as well

Reply:
Sorry if I my previous message was confusing. I should clarify:

I definitely want to keep supporting people who are using F* on .NET. However, as I mentioned before, I no longer want to support building fstar.exe as a .NET executable. So, I'm changing things around in that context. I would still definitely keep ulibfs etc

So, my plan is to just to use our current makefiles to extract all the relevant files from .fst to ulib/fs/extracted/*.fs
 and then just use msbuild with ulibfs.fsproj
 rather than trying to have msbuild call fstar.exe via fsharp.extraction.targets

Reply:
Some confusion which I have, is not from your explanation, but my holes in understanding. I'm mostly interested in F* on .NET and want to have nice IDE experience before I try to play more with F* itself.

From what I see `fsharp.extraction.targets` is nice baseline for that. I was trying to tell my ideas here https://github.com/FStarLang/FStar/pull/2308/files#r837008754 

> So, my plan is to just to use our current makefiles to extract all the relevant files from .fst to ulib/fs/extracted/*.fs
and then just use msbuild with ulibfs.fsproj
rather than trying to have msbuild call fstar.exe via fsharp.extraction.targets

If by a chance I can convince you to replace makefiles with `dotnet` executable for that specific part. Honestly I do not care about donet. just something which is not hostile to Windows developer, like Cygwin. That's sounds like holywar, but I'm looking to find something which is win-win for both communities, and personally for main developers. I relatively familiar with Linux, even if primary I'm Windows dev, but I do develop on Gentoo, so this is not quite regular Linux distro.

My understanding that `fsharp.extraction.targets` also provide experience for using F* within Visual Studio.I do not sure about your plans on that. If you drop that, what would be substitution for that experience?

Also I would like to ask for what style of contribution do you prefer. I technically like work in very small pieces like this while moving toward my long-term goal (look comment above), but if you want, I can dump for example whole new-style Dotnet as one piece :smile:

Reply:
@kant2002 I had a little chat with @nikswamy in the meantime and I think I can try to explain two worlds :) 

The idea is to simplify the build complexity in F* itself. Even when fsharp.extraction.targets is removed from the process of building ulibfs it shouldn't be a problem for people simply depending on a final product. Assuming that ulibfs/fstar.exe is nicely packaged as a nuget for easy consumption from F#/F* projects. 

Having said that, projects that build on ulibfs would still benefit from some build automation that fsharp.extraction.targets provide so we get a better experience with vs/dotnet cmdline. 

So, while building ulibfs doesn't have to depend on fsharp.extraction.targets and it could simply use a handcrafted fsproj that works with generated files, it should still eventually be packaged as a nuget somehow. 

That nuget could provide
- fstar.exe and dependencies (for extraction of fst files in users code)
- ulibfs.dll (needed for runtime of users code)
- targets that provide basic integration / automation with dotnet / Visual Studio / msbuild

Hope that clarifies things a bit.

Reply:
> The idea is to simplify the build complexity in F* itself. Even when fsharp.extraction.targets is removed from the process of building ulibfs it shouldn't be a problem for people simply depending on a final product. 

If I understand correctly, that's mean that `fsharp.extraction.targets` is primarily used for build infra, and that it is used in `examples` folder that is happy coincidence. Correct?

In general seems to be this change goes nowhere, so if that's the case, you can close this PR. 

It seems it would be better for me concentrate on nuget package. I may make this in separate project and show you the results.

Reply:
The world looked differently at the time I was adding the fsharp.extraction.targets :) I wouldn't call it coincidence, but perhaps not the best design choice I've made. 

I think it would be best if we, as a .net based community focused on providing a packaging of F* as a nuget and improving our integration / tooling. 
I think my old PR could serve as a pretty good baseline for doing that work as I've successfully used it in a fairly big codebase last year, but I'm happy to help if you would prefer to do something differently from scratch.